src/webgl/p5.RendererGL.js

```diff
@@ -970,32 +991,92 @@ p5.RendererGL = class RendererGL extends p5.Renderer {
+    let operation = undefined;
     if (typeof args[0] === 'string') {
-      // TODO, handle filter constants:
-      //   this.filterShader = map(args[0], {GRAYSCALE: grayscaleShader, ...})
-      //   filterOperationParameter = undefined or args[1]
-      p5._friendlyError('webgl filter implementation in progress');
-      return;
+      operation = args[0];
+      let defaults = {
+        [constants.BLUR]: 3,
+        [constants.POSTERIZE]: 4,
+        [constants.THRESHOLD]: 0.5
+      };
+      let useDefaultParam = operation in defaults && args[1] === undefined;
+      filterParameter = useDefaultParam ? defaults[operation] : args[1];
+
+      // Create and store shader for constants once on initial filter call.
+      // Need to store multiple in case user calls different filters,
+      // eg. filter(BLUR) then filter(GRAY)
+      if ( !(operation in this.defaultFilterShaders) ) {
+        this.defaultFilterShaders[operation] = new p5.Shader(
+          pg._renderer,
+          filterShaderVert,
+          filterShaderFrags[operation]
+        );
+      }
+      this.filterShader = this.defaultFilterShaders[operation];
     }
-    let userShader = args[0];
-
-    // Copy the user shader once on the initial filter call,
-    // since it has to be bound to pg and not main
-    let isSameUserShader = (
-      this.filterShader !== undefined &&
-      userShader._vertSrc === this.filterShader._vertSrc &&
-      userShader._fragSrc === this.filterShader._fragSrc
-    );
-    if (!isSameUserShader) {
-      this.filterShader =
-        new p5.Shader(pg._renderer, userShader._vertSrc, userShader._fragSrc);
-      this.filterShader.parentShader = userShader;
+    // use custom user-supplied shader
+    else {
+      let userShader = args[0];
+
+      // Copy the user shader once on the initial filter call,
+      // since it has to be bound to pg and not main
+      let isSameUserShader = (
+        this.filterShader !== undefined &&
+        userShader._vertSrc === this.filterShader._vertSrc &&
+        userShader._fragSrc === this.filterShader._fragSrc
+      );
+      if (!isSameUserShader) {
+        this.filterShader =
+          new p5.Shader(pg._renderer, userShader._vertSrc, userShader._fragSrc);
+        this.filterShader.parentShader = userShader;
+      }
     }
 
-    // apply shader to pg
-    pg.shader(this.filterShader);
-    this.filterShader.setUniform('tex0', this);
-    pg.rect(0,0,this.width,this.height);
+    pg.clear(); // prevent undesirable feedback effects accumulating secretly
+
+    // apply blur shader with multiple passes
+    if (operation === constants.BLUR) {
+
+      // initial binding and setup
+      pg.shader(this.filterShader);
+      this.filterShader.setUniform('texelSize', [1/this.width, 1/this.height]);
+
+      // do initial horizontal and vertical pass,
+      // starting with parent renderer as tex0 uniform
+      this.filterShader.setUniform('tex0', this);     // vertically flips first
```

Comment:
why do we need to do these initial passes? I think we could get away with just drawing the source canvas into the pg layer without a shader, and then running the loop.